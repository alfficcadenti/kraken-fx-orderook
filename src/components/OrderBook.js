/* eslint-disable no-mixed-spaces-and-tabs */
import React, { useState, useEffect } from "react";

const OrderBook = () => {
	const [buy, setBuy] = useState([]);
	const [sell, setSell] = useState([]);

	useEffect(() => {
		const subscribe = {
			event: "subscribe",
			feed: "book",
			product_ids: ["PI_XBTUSD"]
		};
		const ws = new WebSocket("wss://futures.kraken.com/ws/v1");
		ws.onopen = () => {
			ws.send(JSON.stringify(subscribe));
		};
		ws.onmessage = (event) => {
			const data = JSON.parse(event.data);
			if (data && data.feed) {
				if (data.side === "buy") {
					const idx = buy.findIndex(x=>x.price === data.price);
					if(idx === -1) {
						setBuy([...buy,{price: data.price, qty: data.qty}]);
					} else {
						const newBuy = buy;
						data.qty ? 
							newBuy[idx] = {price: data.price, qty: data.qty} : 
							newBuy.splice(idx, 1);
						newBuy[idx] = {price: data.price, qty: data.qty};
						setBuy(newBuy);
					}

					

				} else if (data.side === "sell") {
					const idx = sell.findIndex(x=>x.price === data.price);
					if(idx === -1) {
						setSell([...sell,{price: data.price, qty: data.qty}]);
					} else {
						const newSell = sell;
						data.qty ? 
							newSell[idx] = {price: data.price, qty: data.qty} : 
							newSell.splice(idx, 1);
						setSell(newSell);
					}
				}
			}
		};
		ws.onclose = () => {
			ws.close();
		};

		return () => {
			ws.close();
		};
	}, [buy,sell]);

	const orderRows = (arr, side) => {
		if(side === "sell") {
			return arr && arr
				.sort((a, b) => (a.price - b.price))
				.map((item, index) => (item.qty ? order(item, index, side) : ""));
		} else if (side === "buy") {
			return arr && arr
				.sort((a, b) => (b.price - a.price))
				.map((item, index) => (item.qty ? order(item, index, side) : ""));
		}
	};
		

	const order = (order,idx, side) => (
		<tr key={idx}>
			<td> {side === "sell" ? order.price : order.qty} </td>
			<td> {side === "sell" ? order.qty  : order.price} </td>
		</tr>);

	const orderHead = (side) => (
		<thead>
			<tr>
				<th colSpan="2">{side}</th>
			</tr>
			<tr>
				<th>{side === "buy" ? "Q.ty (XBT)" : "Price (USD)"}</th>
				<th>{side === "buy" ? "Price (USD)" : "Q.ty (XBT)"}</th>
			</tr>
		</thead>
	);
 
	return (
		<div className="order-book">
			<div className="order-container">
				<table>
					{orderHead("buy")}
					<tbody>{orderRows(buy, "buy")}</tbody>
				</table>

				<table>
					{orderHead("sell")}
					<tbody>{orderRows(sell, "sell")}</tbody>
				</table>
			</div>
		</div>
	);
};

export default OrderBook;
