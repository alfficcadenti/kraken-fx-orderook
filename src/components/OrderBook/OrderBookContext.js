import React, { useState, createContext } from "react";
import PropTypes from "prop-types";

export const OrderBookContext = createContext([{}, () => {}]);

export const OrderBookProvider = (props) => {
	const [state, setState] = useState({});
	return (
		<OrderBookContext.Provider value={[state, setState]}>
			{props.children}
		</OrderBookContext.Provider>
	);
};

OrderBookProvider.propTypes = { 
	children: PropTypes.node.isRequired
};