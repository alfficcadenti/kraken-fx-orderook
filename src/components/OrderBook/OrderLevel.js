import React, { Fragment } from "react";
import PropTypes from "prop-types"; 

const OrderLevel = (price, qty, total, idx, type, onPriceClick) => (
	<tr key={idx}>
		{type === "buy" ? (<Fragment><td>{total}</td><td>{qty}</td></Fragment>) : <td className="order-price sell" onClick={()=>onPriceClick(price)}>{price}</td>} 
		{type === "buy" ? <td className="order-price buy" onClick={()=>onPriceClick(price)}>{price}</td> : (<Fragment><td>{qty}</td><td>{total}</td></Fragment>)}
	</tr>
);

OrderLevel.propTypes = {
	price: PropTypes.number.isRequired,
	qty: PropTypes.number.isRequired,
	total: PropTypes.number.isRequired,
	idx: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
	type: PropTypes.string.isRequired,
	onPriceClick: PropTypes.func.isRequired,
};

export default OrderLevel;