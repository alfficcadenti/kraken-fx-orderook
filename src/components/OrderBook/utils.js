const volumeAccumulator = (accumulator, curr) => accumulator + curr.qty;

export const generateArrayToDisplay = (arr, side, depthLevel) => {
	const sortedArray = sortOrderBookData(arr, side).slice(0,depthLevel);
	return sortedArray && sortedArray.slice(0,depthLevel)
		.map((x,idx)=>(
			{
				price: x.price, 
				qty: x.qty, 
				total: sortedArray.slice(0,idx+1).reduce(volumeAccumulator,0)
			}
		));
};

export const sortOrderBookData = (arr = [], side = "") => {
	if(side === "sell") {
		return Array.isArray(arr) && arr.sort((a, b) => (a.price - b.price));
	}
	if(side === "buy") {
		return Array.isArray(arr) && arr.sort((a, b) => (b.price - a.price));
	}
	return arr;
};