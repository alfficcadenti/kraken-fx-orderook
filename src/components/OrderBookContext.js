import React, { useState, createContext } from "react";
import PropTypes from "prop-types";

const OrderBookContext = createContext([{}, () => {}]);

const OrderBookProvider = (props) => {
	const [state, setState] = useState({});
	return (
		<OrderBookContext.Provider value={[state, setState]}>
			{props.children}
		</OrderBookContext.Provider>
	);
};

OrderBookProvider.propTypes = { 
	children: PropTypes.node.isRequired 
};

export { OrderBookContext, OrderBookProvider };