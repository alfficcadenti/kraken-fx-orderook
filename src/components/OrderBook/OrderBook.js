import React, { useContext, Fragment } from "react";
import PropTypes from "prop-types";
import { OrderBookContext } from "./OrderBookContext";
import OrderLevel from "./OrderLevel";
import {generateArrayToDisplay} from "../utils/utils";

const OrderBook = ({buy,sell}) => {
	// eslint-disable-next-line no-unused-vars
	const [state, setState] = useContext(OrderBookContext);

	const onPriceClick = (price) => setState(state => ({ ...state, inputPrice: price }));

	const orderRows = (arr, side) => {
		const depthLevel = state.depthLevel || 10;
		const formattedBook = generateArrayToDisplay(arr, side, depthLevel);
		return formattedBook.map((item, index) => (OrderLevel(item.price, item.qty, item.total, index, side, onPriceClick)));
	};

	const orderBookHead = (side) => (
		<thead>
			<tr>
				{side === "buy" ? 
					<Fragment><th>Total</th><th>Q.ty</th><th>Price</th></Fragment> :
					<Fragment><th>Price</th><th>Q.ty</th><th>Total</th></Fragment>
				}
			</tr>
		</thead>
	);
 
	return (
		<div className="order-container">
			<table>
				{orderBookHead("buy")}
				<tbody>{buy.length ? orderRows(buy, "buy") : ""}</tbody>
			</table>

			<table>
				{orderBookHead("sell")}
				<tbody>{sell.length ? orderRows(sell, "sell") : ""}</tbody>
			</table>
		</div>
	);
};

OrderBook.propTypes = {
	buy: PropTypes.array,
	sell: PropTypes.array,
};

export default OrderBook;
