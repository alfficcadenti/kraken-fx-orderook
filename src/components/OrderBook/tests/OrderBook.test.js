import React from "react";
import {shallow} from "enzyme";
import OrderBook from "../OrderBook";
import { OrderBookProvider } from "../OrderBookContext";
import renderer from "react-test-renderer";

const buy = [{price:10, qty: 10}];
const buy2 = [{price:10, qty: 10},{price:1, qty: 12},{price:50, qty: 10}];
const sell = [{price:1, qty: 10}];


describe("<OrderBook>", () => {
	test("renders OrderBook without crashing", () => {
		// eslint-disable-next-line no-unused-vars
		jest.spyOn(React, "useContext").mockImplementation((_context) => "context_value" );
		shallow(<OrderBookProvider><OrderBook buy={buy} sell={sell}/></OrderBookProvider>);
	});
	
	it("matches the snapshot", () => {
		const tree = renderer
			.create(<OrderBook buy={buy} sell={sell}/>)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it("should display total quantity in buy order eql to 32", () => {
		const wrapper = shallow(<OrderBook buy={buy2} sell={sell}/>);
		expect(wrapper.find("table").first()
			.find("tr").last()
			.find("td").first().text()).toEqual("32");
	});
	
	it("should have 2 theads", () => {
		const wrapper = shallow(<OrderBook buy={buy} sell={sell}/>);
		expect(wrapper.find("thead").length).toEqual(2);
	});

	it("should have 2 tbody", () => {
		const wrapper = shallow(<OrderBook buy={buy} sell={sell}/>);
		expect(wrapper.find("tbody").length).toEqual(2);
	});

	it("should have 1 td with class buy and value 10", () => {
		const wrapper = shallow(<OrderBook buy={buy} sell={sell}/>);
		expect(wrapper.find(".buy").text()).toEqual("10");
	});

	it("should have empty tbody when buy and sell are empty array", () => {
		const wrapper = shallow(<OrderBook buy={[]} sell={[]}/>);
		expect(wrapper.find("tbody").length).toEqual(2);
		expect(wrapper.find("tbody").first().text()).toEqual("");
		expect(wrapper.find("tbody").last().text()).toEqual("");
	});
});