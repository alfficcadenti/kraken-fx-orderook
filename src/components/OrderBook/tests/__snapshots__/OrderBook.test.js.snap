// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderBook> matches the snapshot 1`] = `
<div
  className="order-container"
>
  <table>
    <thead>
      <tr>
        <th>
          Total
        </th>
        <th>
          Q.ty
        </th>
        <th>
          Price
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          10
        </td>
        <td>
          10
        </td>
        <td
          className="order-price buy"
          onClick={[Function]}
        >
          10
        </td>
      </tr>
    </tbody>
  </table>
  <table>
    <thead>
      <tr>
        <th>
          Price
        </th>
        <th>
          Q.ty
        </th>
        <th>
          Total
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="order-price sell"
          onClick={[Function]}
        >
          1
        </td>
        <td>
          10
        </td>
        <td>
          10
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
